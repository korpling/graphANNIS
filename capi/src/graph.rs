use super::{cast_const, cstr};
use crate::{cerror::ErrorList, data::IterPtr, map_cerr};
use graphannis::{
    errors::GraphAnnisError,
    graph::{Annotation, Edge, Match, NodeID},
    model::{AnnotationComponent, AnnotationComponentType},
    AnnotationGraph,
};
use itertools::Itertools;
use std::ffi::CString;

/// Get the type of the given component.
#[no_mangle]
pub extern "C" fn annis_component_type(c: *const AnnotationComponent) -> AnnotationComponentType {
    let c: &AnnotationComponent = cast_const(c);
    c.get_type()
}

/// Get the layer of the given component.
///
/// The returned string must be deallocated by the caller using annis_str_free()!
#[no_mangle]
pub extern "C" fn annis_component_layer(c: *const AnnotationComponent) -> *mut libc::c_char {
    let c: &AnnotationComponent = cast_const(c);
    let as_string: &str = &c.layer;
    CString::new(as_string).unwrap_or_default().into_raw()
}

/// Get the name of the given component.
///
/// The returned string must be deallocated by the caller using annis_str_free()!
#[no_mangle]
pub extern "C" fn annis_component_name(c: *const AnnotationComponent) -> *mut libc::c_char {
    let c: &AnnotationComponent = cast_const(c);
    let as_string: &str = &c.name;
    CString::new(as_string).unwrap_or_default().into_raw()
}

/// Return an iterator over all nodes of the graph `g` and the given `node_type` (e.g. "node" or "corpus").
#[no_mangle]
pub extern "C" fn annis_graph_nodes_by_type(
    g: *const AnnotationGraph,
    node_type: *const libc::c_char,
) -> *mut IterPtr<NodeID> {
    let db: &AnnotationGraph = cast_const(g);
    let node_type = cstr(node_type);
    let it = db
        .get_node_annos()
        .exact_anno_search(Some("annis"), "node_type", Some(node_type.as_ref()).into())
        .map_ok(|m: Match| m.node)
        .map(|n| n.map_err(GraphAnnisError::from));
    Box::into_raw(Box::new(Box::new(it)))
}

/// Return a vector of all annotations for the given `node` in the graph `g`.
///
/// - `err` - Pointer to a list of errors. If any error occured, this list will be non-empty.
#[no_mangle]
pub extern "C" fn annis_graph_annotations_for_node(
    g: *const AnnotationGraph,
    node: NodeID,
    err: *mut *mut ErrorList,
) -> *mut Vec<Annotation> {
    let db: &AnnotationGraph = cast_const(g);

    map_cerr(db.get_node_annos().get_annotations_for_item(&node), err)
        .map(|annos| {
            let anno = Box::new(annos);
            Box::into_raw(anno)
        })
        .unwrap_or_else(std::ptr::null_mut)
}

/// Return a vector of all components for the graph `g`.
#[no_mangle]
pub extern "C" fn annis_graph_all_components(
    g: *const AnnotationGraph,
) -> *mut Vec<AnnotationComponent> {
    let db: &AnnotationGraph = cast_const(g);

    Box::into_raw(Box::new(db.get_all_components(None, None)))
}

/// Return a vector of all components for the graph `g` and the given component type.
#[no_mangle]
pub extern "C" fn annis_graph_all_components_by_type(
    g: *const AnnotationGraph,
    ctype: AnnotationComponentType,
) -> *mut Vec<AnnotationComponent> {
    let db: &AnnotationGraph = cast_const(g);

    Box::into_raw(Box::new(db.get_all_components(Some(ctype), None)))
}

/// Return a vector of all outgoing edges for the graph `g`, the `source` node and the given `component`.
///
/// - `err` - Pointer to a list of errors. If any error occured, this list will be non-empty.
#[no_mangle]
pub extern "C" fn annis_graph_outgoing_edges(
    g: *const AnnotationGraph,
    source: NodeID,
    component: *const AnnotationComponent,
    err: *mut *mut ErrorList,
) -> *mut Vec<Edge> {
    let db: &AnnotationGraph = cast_const(g);
    let component: &AnnotationComponent = cast_const(component);

    let mut result: Vec<Edge> = Vec::new();

    if let Some(gs) = db.get_graphstorage(component) {
        for target in gs.get_outgoing_edges(source) {
            if let Some(target) = map_cerr(target, err) {
                result.push(Edge { source, target })
            }
        }
    }

    Box::into_raw(Box::new(result))
}

/// Return a vector of annnotations for the given `edge` in the `component` of graph `g.
///
/// - `err` - Pointer to a list of errors. If any error occured, this list will be non-empty.
#[no_mangle]
pub extern "C" fn annis_graph_annotations_for_edge(
    g: *const AnnotationGraph,
    edge: Edge,
    component: *const AnnotationComponent,
    err: *mut *mut ErrorList,
) -> *mut Vec<Annotation> {
    let db: &AnnotationGraph = cast_const(g);
    let component: &AnnotationComponent = cast_const(component);

    if let Some(gs) = db.get_graphstorage(component) {
        map_cerr(gs.get_anno_storage().get_annotations_for_item(&edge), err)
            .map(|annos| Box::into_raw(Box::new(annos)))
            .unwrap_or_else(std::ptr::null_mut)
    } else {
        Box::into_raw(Box::new(vec![]))
    }
}
