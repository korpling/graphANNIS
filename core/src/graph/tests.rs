use super::*;
use crate::types::{AnnoKey, Annotation, DefaultComponentType, Edge};

#[test]
fn create_writeable_gs() {
    let mut db = Graph::<DefaultComponentType>::new(false).unwrap();

    let anno_key = AnnoKey {
        ns: "test".into(),
        name: "edge_anno".into(),
    };
    let anno_val = "testValue".into();

    let component = Component::new(DefaultComponentType::Edge, "test".into(), "dep".into());
    let gs: &mut dyn WriteableGraphStorage = db.get_or_create_writable(&component).unwrap();

    gs.add_edge(Edge {
        source: 0,
        target: 1,
    })
    .unwrap();

    gs.add_edge_annotation(
        Edge {
            source: 0,
            target: 1,
        },
        Annotation {
            key: anno_key,
            val: anno_val,
        },
    )
    .unwrap();
}

#[test]
fn open_existing_graph_storage() {
    let mut db = Graph::<DefaultComponentType>::new(false).unwrap();

    let component = Component::new(DefaultComponentType::Edge, "test".into(), "dep".into());
    db.get_or_create_writable(&component).unwrap();

    let tmp = tempfile::tempdir().unwrap();

    db.save_to(tmp.path()).unwrap();

    let mut db = Graph::new(false).unwrap();
    db.open(tmp.path()).unwrap();
    assert_eq!(1, db.components.len());
    let gs = db.components.get(&component);
    assert_eq!(true, gs.is_some());
    assert_eq!(false, gs.unwrap().is_some());

    db.ensure_loaded(&component).unwrap();
    assert_eq!(1, db.components.len());
    let gs = db.components.get(&component);
    assert_eq!(true, gs.is_some());
    assert_eq!(true, gs.unwrap().is_some());
}

#[test]
fn load_existing_graph_wit_preload() {
    let mut db = Graph::<DefaultComponentType>::new(false).unwrap();

    let component = Component::new(DefaultComponentType::Edge, "test".into(), "dep".into());
    db.get_or_create_writable(&component).unwrap();

    let tmp = tempfile::tempdir().unwrap();

    db.save_to(tmp.path()).unwrap();

    let mut db = Graph::new(false).unwrap();
    #[allow(deprecated)]
    db.load_from(tmp.path(), true).unwrap();
    assert_eq!(1, db.components.len());
    let gs = db.components.get(&component);
    assert_eq!(true, gs.is_some());
    assert_eq!(true, gs.unwrap().is_some());
}

#[test]
fn load_existing_graph_storage_parallel() {
    let mut db = Graph::<DefaultComponentType>::new(false).unwrap();

    let component = Component::new(DefaultComponentType::Edge, "test".into(), "dep".into());
    db.get_or_create_writable(&component).unwrap();

    let tmp = tempfile::tempdir().unwrap();

    db.save_to(tmp.path()).unwrap();

    let mut db = Graph::new(false).unwrap();
    db.open(tmp.path()).unwrap();
    assert_eq!(1, db.components.len());
    let gs = db.components.get(&component);
    assert_eq!(true, gs.is_some());
    assert_eq!(false, gs.unwrap().is_some());

    db.ensure_loaded_parallel(&[component.clone()]).unwrap();
    assert_eq!(1, db.components.len());
    let gs = db.components.get(&component);
    assert_eq!(true, gs.is_some());
    assert_eq!(true, gs.unwrap().is_some());
}

#[test]
fn load_non_existing_graph_storage_parallel() {
    let mut db = Graph::<DefaultComponentType>::new(false).unwrap();

    let component = Component::new(DefaultComponentType::Edge, "test".into(), "dep".into());

    let tmp = tempfile::tempdir().unwrap();

    db.save_to(tmp.path()).unwrap();

    let mut db = Graph::new(false).unwrap();
    db.open(tmp.path()).unwrap();

    db.ensure_loaded_parallel(&[component]).unwrap();
    assert_eq!(0, db.components.len());
}

#[test]
fn load_with_wal_file() {
    let mut db = Graph::<DefaultComponentType>::new(false).unwrap();
    let example_node = 0;
    db.node_annos
        .insert(
            example_node,
            Annotation {
                key: NODE_TYPE_KEY.as_ref().clone(),
                val: "corpus".into(),
            },
        )
        .unwrap();
    db.node_annos
        .insert(
            example_node,
            Annotation {
                key: NODE_NAME_KEY.as_ref().clone(),
                val: "root".into(),
            },
        )
        .unwrap();

    let tmp = tempfile::tempdir().unwrap();
    // Save and remember the location, so that updates are recorded in a WAL
    // file
    db.persist_to(tmp.path()).unwrap();

    // Add an node annotation with apply_update
    let mut u = GraphUpdate::new();
    u.add_event(UpdateEvent::AddNodeLabel {
        node_name: "root".into(),
        anno_ns: "example".into(),
        anno_name: "anno-name".into(),
        anno_value: "anno-value".into(),
    })
    .unwrap();
    db.apply_update(&mut u, |_| {}).unwrap();

    std::mem::drop(db);

    // Check that loading the database again contains the changes
    let mut db = Graph::<DefaultComponentType>::new(false).unwrap();
    db.open(tmp.path()).unwrap();
    db.ensure_loaded_all().unwrap();
    let anno_value = db
        .node_annos
        .get_value_for_item(
            &example_node,
            &AnnoKey {
                name: "anno-name".into(),
                ns: "example".into(),
            },
        )
        .unwrap()
        .unwrap();
    assert_eq!("anno-value", anno_value);
}

#[test]
fn import_from_existing() {
    let mut other = Graph::<DefaultComponentType>::new(false).unwrap();

    let component = Component::new(DefaultComponentType::Edge, "test".into(), "dep".into());
    other.get_or_create_writable(&component).unwrap();

    let tmp1 = tempfile::tempdir().unwrap();
    let tmp2 = tempfile::tempdir().unwrap();

    other.save_to(tmp1.path()).unwrap();
    other.save_to(tmp2.path()).unwrap();

    // Open the first copy
    let mut db = Graph::new(false).unwrap();
    db.open(tmp1.path()).unwrap();
    assert_eq!(Some(tmp1.path().to_path_buf()), db.location);

    // Import the second location, the location should have been cleared and the
    // graph storages loaded from the second location.
    db.import(tmp2.path()).unwrap();

    assert_eq!(None, db.location);
    assert_eq!(1, db.components.len());

    let gs = db.components.get(&component);
    assert_eq!(true, gs.is_some());
    assert_eq!(true, gs.unwrap().is_some());
}

#[test]
fn update_without_updating_stats() {
    // Create a minimal graph
    let mut db = Graph::<DefaultComponentType>::new(false).unwrap();
    for example_node in 0..10 {
        db.node_annos
            .insert(
                example_node,
                Annotation {
                    key: NODE_TYPE_KEY.as_ref().clone(),
                    val: "corpus".into(),
                },
            )
            .unwrap();
        db.node_annos
            .insert(
                example_node,
                Annotation {
                    key: NODE_NAME_KEY.as_ref().clone(),
                    val: format!("n{example_node}").into(),
                },
            )
            .unwrap();
    }

    db.calculate_all_statistics().unwrap();
    let old_guess = db
        .node_annos
        .guess_max_count(Some("annis"), "node_name", "", &char::MAX.to_string())
        .unwrap();
    assert_eq!(10, old_guess);

    // Add a complete unknown annotation name to the graph
    let mut updates = GraphUpdate::new();
    for example_node in 0..10 {
        updates
            .add_event(UpdateEvent::AddNodeLabel {
                node_name: format!("n{example_node}").into(),
                anno_ns: "test".into(),
                anno_name: "test".into(),
                anno_value: "nothing".into(),
            })
            .unwrap();
    }

    // Apply an update and check that the new annotation layer is still unknown
    // to the statistics
    db.apply_update_keep_statistics(&mut updates, |_| {})
        .unwrap();
    let new_guess = db
        .node_annos
        .guess_max_count(Some("test"), "test", "", &char::MAX.to_string())
        .unwrap();
    assert_eq!(0, new_guess);
}
