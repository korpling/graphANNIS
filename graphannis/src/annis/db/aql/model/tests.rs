use std::{fs::File, path::PathBuf};

use crate::{
    annis::db::{aql::model::CorpusSize, example_generator},
    model::AnnotationComponent,
    AnnotationGraph,
};
use assert_matches::assert_matches;
use graphannis_core::graph::{
    serialization::graphml,
    storage::GraphStorage,
    update::{GraphUpdate, UpdateEvent},
    NODE_NAME_KEY,
};
use insta::assert_snapshot;
use itertools::Itertools;

use super::AnnotationComponentType::{self, Coverage};

#[test]
fn global_stats_token_count() {
    let cargo_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let input_file = File::open(&cargo_dir.join("tests/SegmentationWithGaps.graphml")).unwrap();
    let (graph, _config_str): (AnnotationGraph, _) =
        graphannis_core::graph::serialization::graphml::import(input_file, false, |_status| {})
            .unwrap();

    assert_eq!(true, graph.global_statistics.is_some());
    let global_stats = graph.global_statistics.unwrap();
    assert_eq!(true, global_stats.all_token_in_order_component);
    assert_matches!(global_stats.corpus_size, 
        CorpusSize::Token { base_token_count, segmentation_count }
        if base_token_count == 16
            && segmentation_count.len() == 2
            && *segmentation_count.get("diplomatic").unwrap() == 11 
            && *segmentation_count.get("norm").unwrap() == 13);
}

#[test]
fn inherited_cov_edges_simple_tokenization() {
    // Ad a simple dominance node structure above the example sentence.
    let mut u = GraphUpdate::new();
    example_generator::create_corpus_structure_simple(&mut u);
    example_generator::create_tokens(&mut u, Some("root/doc1"), Some("root/doc1"));
    example_generator::make_span(
        &mut u,
        "root/doc1#span1",
        &["root/doc1#tok1", "root/doc1#tok2", "root/doc1#tok3"],
        true,
    );
    example_generator::make_span(
        &mut u,
        "root/doc1#span2",
        &["root/doc1#tok4", "root/doc1#tok5"],
        true,
    );
    u.add_event(UpdateEvent::AddNode {
        node_name: "root/doc1#struct1".to_string(),
        node_type: "node".to_string(),
    })
    .unwrap();
    u.add_event(UpdateEvent::AddNodeLabel {
        node_name: "root/doc1#struct1".to_string(),
        anno_ns: "test".to_string(),
        anno_name: "cat".to_string(),
        anno_value: "P".to_string(),
    })
    .unwrap();
    u.add_event(UpdateEvent::AddEdge {
        source_node: "root/doc1#struct1".to_string(),
        target_node: "root/doc1#span1".to_string(),
        layer: "test".to_string(),
        component_type: "Dominance".to_string(),
        component_name: "edge".to_string(),
    })
    .unwrap();
    u.add_event(UpdateEvent::AddEdge {
        source_node: "root/doc1#struct1".to_string(),
        target_node: "root/doc1#span2".to_string(),
        layer: "test".to_string(),
        component_type: "Dominance".to_string(),
        component_name: "edge".to_string(),
    })
    .unwrap();
    u.add_event(UpdateEvent::AddNode {
        node_name: "root/doc1#struct2".to_string(),
        node_type: "node".to_string(),
    })
    .unwrap();
    u.add_event(UpdateEvent::AddNodeLabel {
        node_name: "root/doc1#struct2".to_string(),
        anno_ns: "test".to_string(),
        anno_name: "cat".to_string(),
        anno_value: "ROOT".to_string(),
    })
    .unwrap();
    u.add_event(UpdateEvent::AddEdge {
        source_node: "root/doc1#struct2".to_string(),
        target_node: "root/doc1#struct1".to_string(),
        layer: "test".to_string(),
        component_type: "Dominance".to_string(),
        component_name: "edge".to_string(),
    })
    .unwrap();

    let mut g = AnnotationGraph::with_default_graphstorages(false).unwrap();
    g.apply_update(&mut u, |_| {}).unwrap();

    // Check that the inherited coverage edges have been created
    let gs = g
        .get_graphstorage_as_ref(&AnnotationComponent::new(
            Coverage,
            "annis".into(),
            "inherited-coverage".into(),
        ))
        .unwrap();
    let sources: Vec<_> = gs
        .source_nodes()
        .map(|n| {
            g.get_node_annos()
                .get_value_for_item(&n.unwrap(), &NODE_NAME_KEY)
                .unwrap()
                .unwrap()
                .to_string()
        })
        .sorted()
        .collect();
    assert_eq!(sources, vec!["root/doc1#struct1", "root/doc1#struct2"]);

    // Also check that the edges target the right token
    assert_out_edges(
        &g,
        gs,
        "root/doc1#struct1",
        &[
            "root/doc1#tok1",
            "root/doc1#tok2",
            "root/doc1#tok3",
            "root/doc1#tok4",
            "root/doc1#tok5",
        ],
    );
    assert_out_edges(
        &g,
        gs,
        "root/doc1#struct2",
        &[
            "root/doc1#tok1",
            "root/doc1#tok2",
            "root/doc1#tok3",
            "root/doc1#tok4",
            "root/doc1#tok5",
        ],
    );
}

#[test]
fn inherited_cov_edges_multiple_segmentation() {
    let mut u = GraphUpdate::new();
    example_generator::create_corpus_structure_simple(&mut u);
    example_generator::create_multiple_segmentations(&mut u, "root/doc1");
    // Add a simple dominance node structure above the "a" segmentation
    example_generator::make_span(
        &mut u,
        "root/doc1#span1",
        &["root/doc1#a1", "root/doc1#a2", "root/doc1#a3"],
        true,
    );
    example_generator::make_span(&mut u, "root/doc1#span2", &["root/doc1#a4"], true);
    u.add_event(UpdateEvent::AddNode {
        node_name: "root/doc1#struct1".to_string(),
        node_type: "node".to_string(),
    })
    .unwrap();
    u.add_event(UpdateEvent::AddNodeLabel {
        node_name: "root/doc1#struct1".to_string(),
        anno_ns: "test".to_string(),
        anno_name: "cat".to_string(),
        anno_value: "ROOT".to_string(),
    })
    .unwrap();
    u.add_event(UpdateEvent::AddEdge {
        source_node: "root/doc1#struct1".to_string(),
        target_node: "root/doc1#span1".to_string(),
        layer: "test".to_string(),
        component_type: "Dominance".to_string(),
        component_name: "edge".to_string(),
    })
    .unwrap();
    u.add_event(UpdateEvent::AddEdge {
        source_node: "root/doc1#struct1".to_string(),
        target_node: "root/doc1#span2".to_string(),
        layer: "test".to_string(),
        component_type: "Dominance".to_string(),
        component_name: "edge".to_string(),
    })
    .unwrap();

    let mut g = AnnotationGraph::with_default_graphstorages(false).unwrap();
    g.apply_update(&mut u, |_| {}).unwrap();

    // TODO Check that the inherited coverage edges have been created
    let gs = g
        .get_graphstorage_as_ref(&AnnotationComponent::new(
            Coverage,
            "annis".into(),
            "inherited-coverage".into(),
        ))
        .unwrap();

    let sources: Vec<_> = gs
        .source_nodes()
        .map(|n| {
            g.get_node_annos()
                .get_value_for_item(&n.unwrap(), &NODE_NAME_KEY)
                .unwrap()
                .unwrap()
                .to_string()
        })
        .sorted()
        .collect();
    assert_eq!(
        sources,
        vec!["root/doc1#span1", "root/doc1#span2", "root/doc1#struct1"]
    );

    // Also check that the edges target the right timeline items (and not the segmentation nodes)
    assert_out_edges(
        &g,
        gs,
        "root/doc1#span1",
        &[
            "root/doc1#tli1",
            "root/doc1#tli2",
            "root/doc1#tli3",
            "root/doc1#tli4",
        ],
    );
    assert_out_edges(&g, gs, "root/doc1#span2", &["root/doc1#tli5"]);
    assert_out_edges(
        &g,
        gs,
        "root/doc1#struct1",
        &[
            "root/doc1#tli1",
            "root/doc1#tli2",
            "root/doc1#tli3",
            "root/doc1#tli4",
            "root/doc1#tli5",
        ],
    );
}

fn assert_out_edges(
    graph: &AnnotationGraph,
    gs: &dyn GraphStorage,
    source: &str,
    expected: &[&str],
) {
    let out: Vec<_> = gs
        .get_outgoing_edges(
            graph
                .get_node_annos()
                .get_node_id_from_name(source)
                .unwrap()
                .unwrap(),
        )
        .map(|t| {
            graph
                .get_node_annos()
                .get_value_for_item(&t.unwrap(), &NODE_NAME_KEY)
                .unwrap()
                .unwrap()
                .to_string()
        })
        .collect();
    assert_eq!(out, expected);
}
#[test]
fn add_token_to_single_sentence() {
    let content = &include_bytes!("../../../../../tests/single_sentence.graphml")[..];
    let (mut graph, _config) =
        graphml::import::<AnnotationComponentType, _, _>(content, false, |_| {}).unwrap();
    // Create updates that add a new token
    let mut updates = GraphUpdate::new();
    updates
        .add_event(UpdateEvent::AddNode {
            node_name: "single_sentence/zossen#newToken".into(),
            node_type: "node".into(),
        })
        .unwrap();
    updates
        .add_event(UpdateEvent::AddNodeLabel {
            node_name: "single_sentence/zossen#newToken".into(),
            anno_ns: "annis".into(),
            anno_name: "tok".into(),
            anno_value: "".into(),
        })
        .unwrap();
    updates
        .add_event(UpdateEvent::AddEdge {
            source_node: "single_sentence/zossen#newToken".into(),
            target_node: "single_sentence/zossen#text".into(),
            layer: "annis".into(),
            component_type: "PartOf".into(),
            component_name: "".into(),
        })
        .unwrap();
    updates
        .add_event(UpdateEvent::DeleteEdge {
            source_node: "single_sentence/zossen#t4".into(),
            target_node: "single_sentence/zossen#t5".into(),
            layer: "annis".into(),
            component_type: "Ordering".into(),
            component_name: "".into(),
        })
        .unwrap();
    updates
        .add_event(UpdateEvent::AddEdge {
            source_node: "single_sentence/zossen#t4".into(),
            target_node: "single_sentence/zossen#newToken".into(),
            layer: "annis".into(),
            component_type: "Ordering".into(),
            component_name: "".into(),
        })
        .unwrap();
    updates
        .add_event(UpdateEvent::AddEdge {
            source_node: "single_sentence/zossen#newToken".into(),
            target_node: "single_sentence/zossen#t5".into(),
            layer: "annis".into(),
            component_type: "Ordering".into(),
            component_name: "".into(),
        })
        .unwrap();
    graph.apply_update(&mut updates, |_| {}).unwrap();

    let mut output = Vec::<u8>::default();
    graphannis_core::graph::serialization::graphml::export_stable_order(
        &graph,
        None,
        &mut output,
        |_| {},
    )
    .unwrap();
    assert_snapshot!(String::from_utf8_lossy(&output));
}
