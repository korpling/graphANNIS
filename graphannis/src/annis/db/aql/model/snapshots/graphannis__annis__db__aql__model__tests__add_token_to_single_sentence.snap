---
source: graphannis/src/annis/db/aql/model/tests.rs
expression: "String::from_utf8_lossy(&output)"
---
<?xml version="1.0" encoding="UTF-8"?>
<graphml>
    <key id="k0" for="node" attr.name="syntax::cat" attr.type="string"/>
    <key id="k1" for="node" attr.name="annis::doc" attr.type="string"/>
    <key id="k2" for="node" attr.name="annis::layer" attr.type="string"/>
    <key id="k3" for="node" attr.name="annis::node_type" attr.type="string"/>
    <key id="k4" for="node" attr.name="default_ns::pos" attr.type="string"/>
    <key id="k5" for="node" attr.name="annis::tok" attr.type="string"/>
    <key id="k6" for="node" attr.name="annis::tok-whitespace-before" attr.type="string"/>
    <graph edgedefault="directed" parse.order="nodesfirst" parse.nodeids="free" parse.edgeids="canonical">
        <node id="single_sentence">
            <data key="k3">corpus</data>
        </node>
        <node id="single_sentence/zossen">
            <data key="k1">zossen</data>
            <data key="k3">corpus</data>
        </node>
        <node id="single_sentence/zossen#text">
            <data key="k3">datasource</data>
        </node>
        <node id="single_sentence/zossen#n1">
            <data key="k0">ROOT</data>
            <data key="k2">syntax</data>
            <data key="k3">node</data>
        </node>
        <node id="single_sentence/zossen#n2">
            <data key="k0">S</data>
            <data key="k2">syntax</data>
            <data key="k3">node</data>
        </node>
        <node id="single_sentence/zossen#n3">
            <data key="k0">NP</data>
            <data key="k2">syntax</data>
            <data key="k3">node</data>
        </node>
        <node id="single_sentence/zossen#t1">
            <data key="k2">default_layer</data>
            <data key="k3">node</data>
            <data key="k4">ART</data>
            <data key="k5">Die</data>
        </node>
        <node id="single_sentence/zossen#t2">
            <data key="k2">default_layer</data>
            <data key="k3">node</data>
            <data key="k4">NN</data>
            <data key="k5">Jugendlichen</data>
            <data key="k6"> </data>
        </node>
        <node id="single_sentence/zossen#n4">
            <data key="k0">PP</data>
            <data key="k2">syntax</data>
            <data key="k3">node</data>
        </node>
        <node id="single_sentence/zossen#t3">
            <data key="k2">default_layer</data>
            <data key="k3">node</data>
            <data key="k4">APPR</data>
            <data key="k5">in</data>
            <data key="k6"> </data>
        </node>
        <node id="single_sentence/zossen#t4">
            <data key="k2">default_layer</data>
            <data key="k3">node</data>
            <data key="k4">NE</data>
            <data key="k5">Zossen</data>
            <data key="k6"> </data>
        </node>
        <node id="single_sentence/zossen#t5">
            <data key="k2">default_layer</data>
            <data key="k3">node</data>
            <data key="k4">VMFIN</data>
            <data key="k5">wollen</data>
            <data key="k6"> </data>
        </node>
        <node id="single_sentence/zossen#n5">
            <data key="k0">NP</data>
            <data key="k2">syntax</data>
            <data key="k3">node</data>
        </node>
        <node id="single_sentence/zossen#t6">
            <data key="k2">default_layer</data>
            <data key="k3">node</data>
            <data key="k4">ART</data>
            <data key="k5">ein</data>
            <data key="k6"> </data>
        </node>
        <node id="single_sentence/zossen#t7">
            <data key="k2">default_layer</data>
            <data key="k3">node</data>
            <data key="k4">NN</data>
            <data key="k5">Musikcaf√©</data>
            <data key="k6"> </data>
        </node>
        <node id="single_sentence/zossen#t8">
            <data key="k2">default_layer</data>
            <data key="k3">node</data>
            <data key="k4">$.</data>
            <data key="k5">.</data>
            <data key="k6"> </data>
        </node>
        <node id="single_sentence/zossen#newToken">
            <data key="k3">node</data>
            <data key="k5"></data>
        </node>
        <edge id="e0" source="single_sentence/zossen#n1" target="single_sentence/zossen#n2" label="Dominance/syntax/">
        </edge>
        <edge id="e1" source="single_sentence/zossen#n1" target="single_sentence/zossen#t8" label="Dominance/syntax/">
        </edge>
        <edge id="e2" source="single_sentence/zossen#n2" target="single_sentence/zossen#n3" label="Dominance/syntax/">
        </edge>
        <edge id="e3" source="single_sentence/zossen#n2" target="single_sentence/zossen#t5" label="Dominance/syntax/">
        </edge>
        <edge id="e4" source="single_sentence/zossen#n2" target="single_sentence/zossen#n5" label="Dominance/syntax/">
        </edge>
        <edge id="e5" source="single_sentence/zossen#n3" target="single_sentence/zossen#t1" label="Dominance/syntax/">
        </edge>
        <edge id="e6" source="single_sentence/zossen#n3" target="single_sentence/zossen#t2" label="Dominance/syntax/">
        </edge>
        <edge id="e7" source="single_sentence/zossen#n3" target="single_sentence/zossen#n4" label="Dominance/syntax/">
        </edge>
        <edge id="e8" source="single_sentence/zossen#n4" target="single_sentence/zossen#t3" label="Dominance/syntax/">
        </edge>
        <edge id="e9" source="single_sentence/zossen#n4" target="single_sentence/zossen#t4" label="Dominance/syntax/">
        </edge>
        <edge id="e10" source="single_sentence/zossen#n5" target="single_sentence/zossen#t6" label="Dominance/syntax/">
        </edge>
        <edge id="e11" source="single_sentence/zossen#n5" target="single_sentence/zossen#t7" label="Dominance/syntax/">
        </edge>
        <edge id="e12" source="single_sentence/zossen#n1" target="single_sentence/zossen#n2" label="Dominance/syntax/edge">
        </edge>
        <edge id="e13" source="single_sentence/zossen#n1" target="single_sentence/zossen#t8" label="Dominance/syntax/edge">
        </edge>
        <edge id="e14" source="single_sentence/zossen#n2" target="single_sentence/zossen#n3" label="Dominance/syntax/edge">
        </edge>
        <edge id="e15" source="single_sentence/zossen#n2" target="single_sentence/zossen#t5" label="Dominance/syntax/edge">
        </edge>
        <edge id="e16" source="single_sentence/zossen#n2" target="single_sentence/zossen#n5" label="Dominance/syntax/edge">
        </edge>
        <edge id="e17" source="single_sentence/zossen#n3" target="single_sentence/zossen#t1" label="Dominance/syntax/edge">
        </edge>
        <edge id="e18" source="single_sentence/zossen#n3" target="single_sentence/zossen#t2" label="Dominance/syntax/edge">
        </edge>
        <edge id="e19" source="single_sentence/zossen#n3" target="single_sentence/zossen#n4" label="Dominance/syntax/edge">
        </edge>
        <edge id="e20" source="single_sentence/zossen#n4" target="single_sentence/zossen#t3" label="Dominance/syntax/edge">
        </edge>
        <edge id="e21" source="single_sentence/zossen#n4" target="single_sentence/zossen#t4" label="Dominance/syntax/edge">
        </edge>
        <edge id="e22" source="single_sentence/zossen#n5" target="single_sentence/zossen#t6" label="Dominance/syntax/edge">
        </edge>
        <edge id="e23" source="single_sentence/zossen#n5" target="single_sentence/zossen#t7" label="Dominance/syntax/edge">
        </edge>
        <edge id="e24" source="single_sentence/zossen#t1" target="single_sentence/zossen#t2" label="Ordering/annis/">
        </edge>
        <edge id="e25" source="single_sentence/zossen#t2" target="single_sentence/zossen#t3" label="Ordering/annis/">
        </edge>
        <edge id="e26" source="single_sentence/zossen#t3" target="single_sentence/zossen#t4" label="Ordering/annis/">
        </edge>
        <edge id="e27" source="single_sentence/zossen#t4" target="single_sentence/zossen#newToken" label="Ordering/annis/">
        </edge>
        <edge id="e28" source="single_sentence/zossen#t5" target="single_sentence/zossen#t6" label="Ordering/annis/">
        </edge>
        <edge id="e29" source="single_sentence/zossen#t6" target="single_sentence/zossen#t7" label="Ordering/annis/">
        </edge>
        <edge id="e30" source="single_sentence/zossen#t7" target="single_sentence/zossen#t8" label="Ordering/annis/">
        </edge>
        <edge id="e31" source="single_sentence/zossen#newToken" target="single_sentence/zossen#t5" label="Ordering/annis/">
        </edge>
        <edge id="e32" source="single_sentence/zossen" target="single_sentence" label="PartOf/annis/">
        </edge>
        <edge id="e33" source="single_sentence/zossen#text" target="single_sentence/zossen" label="PartOf/annis/">
        </edge>
        <edge id="e34" source="single_sentence/zossen#n1" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e35" source="single_sentence/zossen#n2" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e36" source="single_sentence/zossen#n3" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e37" source="single_sentence/zossen#t1" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e38" source="single_sentence/zossen#t2" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e39" source="single_sentence/zossen#n4" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e40" source="single_sentence/zossen#t3" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e41" source="single_sentence/zossen#t4" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e42" source="single_sentence/zossen#t5" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e43" source="single_sentence/zossen#n5" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e44" source="single_sentence/zossen#t6" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e45" source="single_sentence/zossen#t7" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e46" source="single_sentence/zossen#t8" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
        <edge id="e47" source="single_sentence/zossen#newToken" target="single_sentence/zossen#text" label="PartOf/annis/">
        </edge>
    </graph>
</graphml>
